glean-core/ios/Glean/Metrics/LabeledMetric.swift

```diff
@@ -0,0 +1,122 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+typealias LabeledMetricCtor = (FfiStr?, FfiStr?, RawStringArray?, Int32, Int32, UInt8, RawStringArray?, Int32) -> UInt64
+
+/// This implements the developer facing API for labeled metrics.
+///
+/// Instances of this class type are automatically generated by the parsers at build time,
+/// allowing developers to record values that were previously registered in the metrics.yaml file.
+///
+/// Unlike most metric types, LabeledMetricType does not have its own corresponding storage engine,
+/// but records metrics for the underlying metric type `T` in the storage engine for that type.
+/// The only difference is that labeled metrics are stored with the special key `$category.$name/$label`.
+public class LabeledMetricType<T> {
+    let handle: UInt64
+    let disabled: Bool
+    let sendInPings: [String]
+    let labels: Set<String>?
+    let subMetric: T
+
+    /// The public constructor used by automatically generated metrics.
+    public init(
+        category: String,
+        name: String,
+        sendInPings: [String],
+        lifetime: Lifetime,
+        disabled: Bool,
+        subMetric: T,
+        labels: Set<String>? = nil
+    ) throws {
+        self.disabled = disabled
+        self.sendInPings = sendInPings
+        self.labels = labels
+        self.subMetric = subMetric
+
+        var metricTypeInstantiator: LabeledMetricCtor
+        switch subMetric {
+        case is CounterMetricType:
+            metricTypeInstantiator = glean_new_labeled_counter_metric
+        // TODO(bug XYZ): Requires the BooleanMetricType first
+        // case is BooleanMetricType:
+        //     metricTypeInstantiator = glean_new_labeled_boolean_metric
+        case is StringMetricType:
+            metricTypeInstantiator = glean_new_labeled_string_metric
+        default:
+            throw "Can not create a labeled version of this metric type"
+        }
+
+        self.handle = withArrayOfCStrings(sendInPings) { pingArray in
+            withArrayOfCStrings(labels?.sorted()) { labels in
+                metricTypeInstantiator(
+                    category,
+                    name,
+                    pingArray,
+                    Int32(sendInPings.count),
+                    lifetime.rawValue,
+                    disabled.toByte(),
+                    labels,
+                    Int32(labels?.count ?? 0)
+                )
+            }
+        }
+    }
+
+    /// Destroy this metric.
+    deinit {
+        if self.handle != 0 {
+            switch self.subMetric {
+            case is CounterMetricType:
+                glean_destroy_labeled_counter_metric(self.handle)
+            // TODO(bug XYZ): Requires the BooleanMetricType first
```

Comment:
I think you filed a [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1583772) for the BooleanMetricType, did you mean to refer to it here?

Reply:
I even changed it locally.... and never committed.